// mui-chips-input@2.0.0/dist/mui-chips-input.es.js
import C, { createElement as me } from "react";
import Q from "@mui/material/Chip";
import { styled as B } from "@mui/material/styles";
import { jsx as x } from "react/jsx-runtime";
import xe from "@mui/icons-material/Close";
import ge from "@mui/material/ClickAwayListener";
import ye from "@mui/material/IconButton";
import be from "@mui/material/TextField";
const O = {
  enter: "Enter",
  backspace: "Backspace"
}, Ee = {
  ime: 229
}, De = B(Q)(({ theme: t, size: n }) => `
    max-width: 100%;
    margin: 2px 4px;
    height: ${n === "small" ? "26px" : "32px"};


    &[aria-disabled="true"] > svg {
      color: ${t.palette.action.disabled};
      cursor: default;
    }

    &.MuiChipsInput-Chip-Editing {
      background-color: ${t.palette.primary.light};
      color: ${t.palette.primary.contrastText};
    }
  `), Ae = {
  ChipStyled: De
}, H = (t) => {
  const {
    className: n,
    index: i,
    onDelete: r,
    disabled: s,
    onEdit: g,
    isEditing: D,
    disableEdition: h,
    ...y
  } = t, c = (d) => {
    d.key === O.enter && r(i);
  }, A = (d) => {
    d?.preventDefault?.(), d?.stopPropagation?.(), r(i);
  }, b = (d) => {
    d.target.textContent === y.label && (s || g(i));
  };
  return /* @__PURE__ */ x(Ae.ChipStyled, {
    className: `MuiChipsInput-Chip ${D ? "MuiChipsInput-Chip-Editing" : ""} ${n || ""}`,
    onKeyDown: c,
    disabled: s,
    onDoubleClick: h ? void 0 : b,
    tabIndex: s ? -1 : 0,
    "aria-disabled": s,
    onDelete: A,
    ...y
  });
};
function Ie(t) {
  return typeof t == "boolean";
}
function we(t) {
  return typeof t == "object" && !Array.isArray(t) && t !== null;
}
function ke(t, n) {
  typeof n == "function" ? n(t) : n && we(n) && "current" in n && (n.current = t);
}
const Oe = B("div")`
  top: 50%;
  transform: translateY(-50%);
  right: 10px;
  position: absolute;
`, Te = B(be)((t) => `
    max-width: 100%;

    .MuiInputBase-root {
      display: flex;
      flex-wrap: wrap;
      align-items: flex-start;
      row-gap: 5px;
      padding-top: ${t.size === "small" ? "5px" : "9px"};
      padding-right: ${t.InputProps?.endAdornment ? "30px" : "9px"};
      padding-bottom: ${t.size === "small" ? "5px" : "9px"};
      padding-left: 10px;

      input {
        min-width: 30px;
        width: auto;
        flex-grow: 1;
        text-overflow: ellipsis;
        padding: ${t.size === "small" ? "3.5px 4px" : "7.5px 4px"};
        align-self: center;
      }
    }
  `), $e = B(Q)(({ theme: t, size: n }) => `
    max-width: 100%;
    margin: 2px 4px;
    height: ${n === "small" ? "26px" : "32px"};

    &[aria-disabled="true"] > svg.MuiChip-deleteIcon {
      color: ${t.palette.action.disabled};
      cursor: default;
    }
  `), J = {
  ChipStyled: $e,
  TextFieldStyled: Te,
  EndAdornmentClose: Oe
}, X = C.forwardRef((t, n) => {
  const {
    chips: i,
    onAddChip: r,
    onEditChip: s,
    onDeleteChip: g,
    onDeleteAllChips: D,
    InputProps: h,
    onInputChange: y,
    disabled: c,
    clearInputOnBlur: A,
    validate: b,
    error: d,
    helperText: T,
    hideClearAll: I,
    inputProps: P,
    size: $,
    disableDeleteOnBackspace: K,
    disableEdition: V,
    className: M,
    renderChip: v,
    addOnWhichKey: w,
    onFocus: o,
    inputValue: p,
    ...S
  } = t, [Z, ee] = C.useState(""), [z, Y] = C.useState(""), N = C.useRef(null), F = C.useRef(!1), te = C.useRef(typeof p == "string"), [m, W] = C.useState(null), {
    onKeyDown: ne,
    ...ie
  } = P || {}, {
    inputRef: Se,
    ...le
  } = h || {}, j = () => {
    Y("");
  }, _ = te.current, k = _ ? p : Z, R = (e) => {
    y?.(e), _ || ee(e);
  }, re = (e) => {
    R(i[e]), W(e), j();
  }, E = () => {
    W(null);
  }, f = () => {
    j(), R("");
  }, oe = (e) => {
    R(e.target.value);
  }, L = () => {
    !F.current || (m !== null ? (E(), f()) : A && f(), F.current = !1);
  }, ae = (e) => {
    N.current = e, n && ke(e, n);
  }, U = (e, l) => (u) => {
    if (typeof b == "function") {
      const a = b(e);
      if (a === !1) {
        l.preventDefault();
        return;
      }
      if (!Ie(a) && a.isError) {
        l.preventDefault(), Y(a.textError);
        return;
      }
    }
    u();
  }, se = (e, l, u) => {
    U(e, u)(() => {
      s?.(e, l), E(), f();
    });
  }, de = (e, l) => {
    U(e, l)(() => {
      r?.(k.trim()), f();
    });
  }, pe = (e, l) => l === Ee.ime ? !1 : w ? Array.isArray(w) ? w.some((u) => u === e) : w === e : e === O.enter, ue = (e) => {
    const l = pe(e.key, e.keyCode), u = e.key === O.backspace, a = k.trim();
    if (e.code === "Tab") {
      L();
      return;
    }
    if (l && e.preventDefault(), k.length > 0 && l)
      a.length === 0 ? f() : m !== null ? se(a, m, e) : de(a, e);
    else if (u && k.length === 0 && i.length > 0 && !K) {
      const q = i.length - 1;
      g?.(q), m === q && E();
    }
    ne?.(e);
  }, ce = (e) => {
    e.preventDefault(), o?.(e), F.current = !0;
  }, he = (e) => {
    e.preventDefault(), !I && !c && (D?.(), f(), E());
  }, fe = (e) => {
    e === m ? (f(), E()) : re(e), N.current?.focus();
  }, Ce = (e) => {
    c || (g?.(e), m !== null && (E(), f()));
  }, G = i.length > 0;
  return /* @__PURE__ */ x(ge, {
    onClickAway: L,
    children: /* @__PURE__ */ x(J.TextFieldStyled, {
      value: k,
      onChange: oe,
      ref: n,
      className: `MuiChipsInput-TextField ${M || ""}`,
      size: $,
      placeholder: "Type and press enter",
      onFocus: ce,
      inputProps: {
        onKeyDown: ue,
        ...ie
      },
      disabled: c,
      error: Boolean(z) || d,
      helperText: z || T,
      InputProps: {
        inputRef: ae,
        startAdornment: G ? i.map((e, l) => {
          const u = `chip-${l}`, a = {
            index: l,
            onEdit: fe,
            label: e,
            title: e,
            isEditing: l === m,
            size: $,
            disabled: c,
            disableEdition: V,
            onDelete: Ce
          };
          return v ? v(H, u, a) : /* @__PURE__ */ me(H, {
            ...a,
            key: u
          });
        }) : null,
        endAdornment: I ? null : /* @__PURE__ */ x(J.EndAdornmentClose, {
          style: {
            visibility: G ? "visible" : "hidden"
          },
          children: /* @__PURE__ */ x(ye, {
            "aria-label": "Clear",
            title: "Clear",
            disabled: c,
            size: "small",
            onClick: he,
            children: /* @__PURE__ */ x(xe, {
              fontSize: "small"
            })
          })
        }),
        ...le
      },
      ...S
    })
  });
});
X.defaultProps = {
  onInputChange: () => {
  },
  clearInputOnBlur: !1,
  hideClearAll: !1,
  disableDeleteOnBackspace: !1,
  disableEdition: !1,
  addOnWhichKey: O.enter,
  onDeleteChip: () => {
  },
  onAddChip: () => {
  },
  inputValue: void 0,
  onEditChip: () => {
  },
  renderChip: void 0,
  onDeleteAllChips: () => {
  },
  validate: () => !0
};
function Be(t, n) {
  return [...t, n];
}
function Pe(t, n) {
  return t.filter((i, r) => n !== r);
}
function Ke(t, n, i) {
  return t.map((r, s) => n === s ? i : r);
}
const ve = C.forwardRef((t, n) => {
  const {
    value: i,
    onChange: r,
    onAddChip: s,
    onInputChange: g,
    onDeleteChip: D,
    disabled: h,
    validate: y,
    clearInputOnBlur: c,
    hideClearAll: A,
    disableDeleteOnBackspace: b,
    onEditChip: d,
    renderChip: T,
    disableEdition: I,
    addOnWhichKey: P,
    inputValue: $,
    ...K
  } = t;
  return /* @__PURE__ */ x(X, {
    chips: i,
    onAddChip: (o) => {
      if (h)
        return;
      const p = Be(i, o), S = p.length - 1;
      s?.(o, S), r?.(p);
    },
    onInputChange: g,
    disableDeleteOnBackspace: b,
    onDeleteChip: (o) => {
      if (h)
        return;
      const p = i[o];
      r?.(Pe(i, o)), D?.(p, o);
    },
    onEditChip: (o, p) => {
      h || I || (r?.(Ke(i, p, o)), d?.(o, p));
    },
    renderChip: T,
    onDeleteAllChips: () => {
      r?.([]);
    },
    clearInputOnBlur: c,
    disabled: h,
    disableEdition: I,
    validate: y,
    inputValue: $,
    hideClearAll: A,
    addOnWhichKey: P,
    ...K,
    ref: n
  });
});
ve.defaultProps = {
  value: [],
  onChange: () => {
  },
  onAddChip: () => {
  },
  onDeleteChip: () => {
  },
  onInputChange: () => {
  },
  onEditChip: () => {
  },
  addOnWhichKey: O.enter,
  clearInputOnBlur: !1,
  disableEdition: !1,
  hideClearAll: !1,
  disableDeleteOnBackspace: !1,
  validate: () => !0
};
export {
  ve as MuiChipsInput
};
